fd/jacobian.go

```diff
@@ -0,0 +1,204 @@
+//      [ ∂f_m/∂x_1 ... ∂f_m/∂x_n ]
+// or element-wise
+//  J_ij = ∂f_i/∂x_j.
+//
+// If dst is not nil, the result will be stored in-place into dst and returned,
+// otherwise a new matrix will be allocated first. Finite difference formula and
+// other options are specified by settings. If settings is nil, the Jacobian
+// will be estimated using the Forward formula and a default step size.
+//
+// Jacobian panics if dst is not nil and its size is not m × len(x), or if the
+// derivative order of the formula is not 1.
+func Jacobian(dst *mat64.Dense, f func(y, x []float64), m int, x []float64, settings *JacobianSettings) *mat64.Dense {
+	n := len(x)
+	if dst == nil {
+		dst = mat64.NewDense(m, n, nil)
+	}
+	r, c := dst.Dims()
+	if r != m || c != n {
+		panic("jacobian: mismatched matrix size")
+	}
+
+	if settings == nil {
+		settings = &JacobianSettings{}
+	}
+	if settings.OriginValue != nil && len(settings.OriginValue) != m {
+		panic("jacobian: mismatched OriginValue slice length")
+	}
+
+	formula := settings.Formula
+	if formula.isZero() {
+		formula = Forward
+	}
+	if formula.Derivative == 0 || formula.Stencil == nil || formula.Step == 0 {
+		panic("jacobian: bad formula")
+	}
+	if formula.Derivative != 1 {
+		panic("jacobian: invalid derivative order")
+	}
+
+	step := settings.Step
+	if step == 0 {
+		step = formula.Step
+	}
+
+	evals := n * len(formula.Stencil)
+	for _, pt := range formula.Stencil {
+		if pt.Loc == 0 {
+			evals -= n - 1
+			break
+		}
+	}
+	nWorkers := 1
+	if settings.Concurrent {
+		nWorkers = runtime.GOMAXPROCS(0)
+		if nWorkers > evals {
+			nWorkers = evals
+		}
+	}
+	if nWorkers == 1 {
+		jacobianSerial(dst, f, x, settings.OriginValue, formula, step)
+	} else {
+		jacobianConcurrent(dst, f, x, settings.OriginValue, formula, step, nWorkers)
+	}
+
+	return dst
+}
+
+func jacobianSerial(dst *mat64.Dense, f func([]float64, []float64), x, origin []float64, formula Formula, step float64) {
+	m, n := dst.Dims()
+	xcopy := make([]float64, n)
+	y := make([]float64, m)
+	col := make([]float64, m)
+	for j := 0; j < n; j++ {
+		for i := range col {
+			col[i] = 0
+		}
+		for _, pt := range formula.Stencil {
+			if pt.Loc == 0 {
+				if origin == nil {
+					origin = make([]float64, m)
+					copy(xcopy, x)
+					f(origin, xcopy)
+				}
+				floats.AddScaled(col, pt.Coeff, origin)
+			} else {
+				copy(xcopy, x)
+				xcopy[j] += pt.Loc * step
+				f(y, xcopy)
+				floats.AddScaled(col, pt.Coeff, y)
+			}
+		}
+		dst.SetCol(j, col)
+	}
+	dst.Scale(1/step, dst)
+}
+
+func jacobianConcurrent(dst *mat64.Dense, f func([]float64, []float64), x, origin []float64, formula Formula, step float64, nWorkers int) {
+	m, n := dst.Dims()
+	for i := 0; i < m; i++ {
+		for j := 0; j < n; j++ {
+			dst.Set(i, j, 0)
+		}
+	}
+
+	var (
+		hasOrigin   bool
+		originCoeff float64
+		once        sync.Once
+		wg          sync.WaitGroup
+	)
+	mus := make([]sync.Mutex, n) // Guard access to individual columns.
+	worker := func(jobs <-chan jacJob) {
```

Comment:
It's cleaner to have a single goroutine that computes the origin value (if necessary) that is under the same waitgroup. The distributor only distributes non-zero locations. This way there is no Once and no if statement.When all goroutines have completed, you know origin is populated, and can do the same thing as you do now.